src/zmq.cpp

```diff
@@ -247,6 +247,16 @@ int zmq_ctx_destroy (void *ctx_)
 
 // Sockets
 
+static zmq::socket_base_t *as_socket_base_t (void *s_)
```

Comment:
should this be a macro instead?

Reply:
What would be the benefit of a macro? Using a function is more readable/debuggable/...

A macro could be used to further reduce the code duplication at the call sites, as it could include the method return. But this might impair readability.

Reply:
Actually I don't really care, it just been the standard in the project so far. I'm good with both.
Anyway two benefits of macro:
1. Performance - those methods are being called a lot (millions of time a second), I'm not sure how the compiler will treat them, we just added a function call. It is probably nothing. however I saw in the past how inlining and macro improve performance.
2. You would'n need the if check on the return value of the function. Just call the MACRO in the beginning of every usage.